### PR TITLE
SONARSLANG-54: Add loops to slang

### PR DESCRIPTION
**Pull request by marcelo-sousa-sonarsource**

### CODE REVIEW

File: slang-api/src/test/java/com/sonarsource/slang/impl/LoopTreeImplTest.java [hunk 1]
```diff
@@ -0,0 +1,70 @@
...
+    assertThat(forTree.children()).containsExactly(condition, body);
+    assertThat(forTree.condition()).isEqualTo(condition);
+    assertThat(forTree.body()).isEqualTo(body);
+    assertThat(forTree.kind()).isEqualTo(FOR);
+
+    LoopTreeImpl whileTree = new LoopTreeImpl(meta, condition, body, WHILE, whileToken);
+    assertThat(whileTree.children()).containsExactly(condition, body);
+    assertThat(whileTree.condition()).isEqualTo(condition);
+    assertThat(whileTree.body()).isEqualTo(body);
+    assertThat(whileTree.kind()).isEqualTo(WHILE);
```
Comment (pynicolas) [inline slang-api/src/test/java/com/sonarsource/slang/impl/LoopTreeImplTest.java]: I don't think that these assertions bring value: they basically check the same thing as lines 45 to 49.

File: sonar-slang-plugin/src/test/java/com/sonarsource/slang/kotlin/KotlinConverterTest.java [hunk 2]
```diff
@@ -340,6 +346,48 @@ public void testComplexMatchExpression() {
...
+    assertTree(forLoop.condition()).isInstanceOf(NativeTree.class);
+    assertThat(forLoop.condition().children()).hasSize(2);
+    assertTree(forLoop.condition().children().get(0)).hasParameterName("item");
+    assertTree(forLoop.condition().children().get(1)).isIdentifier("ints");
+    assertTree(forLoop.body()).isBlock(AssignmentExpressionTree.class, AssignmentExpressionTree.class);
+    assertThat(forLoop.kind()).isEqualTo(FOR);
+    assertThat(forLoop.keyword().text()).isEqualTo("for");
+    assertTree(forLoop).isNotEquivalentTo(kotlinStatement("for (item : String in ints) { x = item; x = x + 1; }"));
+    assertTree(forLoop).isNotEquivalentTo(kotlinStatement("for (it : Int in ints) { x = item; x = x + 1; }"));
+    assertTree(forLoop).isNotEquivalentTo(kotlinStatement("for (item : Int in floats) { x = item; x = x + 1; }"));
```
Comment (pynicolas) [inline sonar-slang-plugin/src/test/java/com/sonarsource/slang/kotlin/KotlinConverterTest.java]: Maybe we could at least one assertion where 2 trees are equivalent. Same thing for the other test methods on loops.

File: sonar-slang-plugin/src/main/java/com/sonarsource/slang/kotlin/KotlinTreeVisitor.java [hunk 3]
```diff
@@ -346,6 +357,41 @@ private Tree createMatchCase(TreeMetaData metaData, KtWhenEntry element) {
...
+    }
+
+    if (ktLoopExpression instanceof KtForExpression) {
+      KtForExpression forExpression = (KtForExpression) ktLoopExpression;
+      Tree parameter = createElement(forExpression.getLoopParameter());
+      Tree range = createElement(forExpression.getLoopRange());
+      if (parameter == null || range == null) {
+        return convertElementToNative(ktLoopExpression, metaData);
+      }
+      Tree condition = createNativeTree(metaData, new KotlinNativeKind(ktLoopExpression), Arrays.asList(parameter, range));
```
Comment (pynicolas) [inline sonar-slang-plugin/src/main/java/com/sonarsource/slang/kotlin/KotlinTreeVisitor.java]: The metadata here is wrong.

File: slang-antlr/src/main/java/com/sonarsource/slang/parser/SLangConverter.java [hunk 4]
```diff
@@ -327,6 +331,29 @@ public Tree visitMatchCase(SLangParser.MatchCaseContext ctx) {
       return new MatchCaseTreeImpl(meta(ctx), expression, body);
     }
 
+    @Override
+    public Tree visitForLoop(SLangParser.ForLoopContext ctx) {
+      Tree declaration = visit(ctx.declaration());
+      Tree statement = visit(ctx.statement());
+      Tree condition = new NativeTreeImpl(meta(ctx), new SNativeKind(ctx), Arrays.asList(declaration, statement));
```
[the remark this answers is not in the document]
Comment (marcelo-sousa-sonarsource) [reply slang-antlr/src/main/java/com/sonarsource/slang/parser/SLangConverter.java]: Fixed in d2aca51 by changing the grammar according to your suggestion.

File: slang-antlr/src/main/antlr4/com/sonarsource/slang/parser/SLang.g4 [hunk 5]
```diff
@@ -168,6 +169,24 @@ matchCase
...
   ;
 
+loopExpression
+  :  forLoop
+  |  whileLoop
+  |  doWhileLoop
+  ;
+
+forLoop
+  :  FOR LPAREN declaration COLON statement RPAREN controlBlock
```
[the remark this answers is not in the document]
Comment (marcelo-sousa-sonarsource) [reply slang-antlr/src/main/antlr4/com/sonarsource/slang/parser/SLang.g4]: I agree; it is a little bit strange to have the declaration as the condition of the loop but I believe it simplifies the code so I have changed it in d2aca51

File: sonar-slang-plugin/src/test/java/com/sonarsource/slang/kotlin/KotlinConverterTest.java (hunk 2 above)
Comment (marcelo-sousa-sonarsource) [reply sonar-slang-plugin/src/test/java/com/sonarsource/slang/kotlin/KotlinConverterTest.java]: Fixed in d2aca51. I have also added some comparison between kotlin and slang statements.

File: sonar-slang-plugin/src/main/java/com/sonarsource/slang/kotlin/KotlinTreeVisitor.java (hunk 3 above)
Comment (marcelo-sousa-sonarsource) [reply sonar-slang-plugin/src/main/java/com/sonarsource/slang/kotlin/KotlinTreeVisitor.java]: Nice catch. Fixed in d2aca51

**Pull request merged**